### ISSUE TITLE
Beacon node can not sync after a hard reset.

### ISSUE THREAD

**Issue by SimplyCorey** [sigp/lighthouse#1616]
## Description

Beacon node won't sync after a hard reset.

## Version

Please provide your Lighthouse and Rust version. Are you building from
`master`, which commit?

lighthouse was built from the `sync-hack` branch.

Lighthouse v0.2.9-b3469c8
BLS Library: blst

## Present Behaviour

The beacon node won't sync after a hard reset.  The hard reset was caused by my docker container crashing.  It jumps between "synced" and "syncing head chain"

To reproduce:
Set up a beacon node and fully sync it.  Have your computer (or container) hard reset.  Observe that the beacon node can not sync anymore.

```
Sep 13 12:52:02.002 INFO Syncing                                 est_time: --, distance: 1112 slots (3 hrs 42 mins), peers: 10, service: slot_notifier
Sep 13 12:52:09.270 WARN Batch failed to download. Dropping chain scoring peers, batch_epoch: 8961, score_adjustment: Low Tolerance Error, chain: 15802250622902401264, service: sync
Sep 13 12:52:09.271 INFO Sync state updated                      new_state: Syncing Head Chain, old_state: Syncing Finalized Chain, service: sync
Sep 13 12:52:13.999 INFO Syncing                                 est_time: --, distance: 1113 slots (3 hrs 42 mins), peers: 10, service: slot_notifier
Sep 13 12:52:25.999 INFO Syncing                                 est_time: --, distance: 1114 slots (3 hrs 42 mins), peers: 10, service: slot_notifier
Sep 13 12:52:30.396 INFO Sync state updated                      new_state: Syncing Finalized Chain, old_state: Syncing Head Chain, service: sync
Sep 13 12:52:37.998 INFO Syncing                                 est_time: --, distance: 1115 slots (3 hrs 43 mins), peers: 12, service: slot_notifier
Sep 13 12:52:49.998 INFO Syncing                                 est_time: --, distance: 1116 slots (3 hrs 43 mins), peers: 12, service: slot_notifier
Sep 13 12:52:51.764 WARN Batch failed to download. Dropping chain scoring peers, batch_epoch: 8961, score_adjustment: Low Tolerance Error, chain: 6772112588494642972, service: sync
Sep 13 12:52:51.765 INFO Sync state updated                      new_state: Syncing Head Chain, old_state: Syncing Finalized Chain, service: sync
Sep 13 12:53:01.813 INFO Sync state updated                      new_state: Synced, old_state: Syncing Head Chain, service: sync
Sep 13 12:53:01.998 INFO Synced                                  slot: 287964, block:    …  empty, epoch: 8998, finalized_epoch: 8961, finalized_root: 0x655b…27d7, peers: 12, service: slot_notifier
Sep 13 12:53:03.532 INFO Sync state updated                      new_state: Syncing Finalized Chain, old_state: Synced, service: sync
Sep 13 12:53:03.781 INFO Sync state updated                      new_state: Synced, old_state: Syncing Finalized Chain, service: sync
Sep 13 12:53:13.997 INFO Synced                                  slot: 287965, block:    …  empty, epoch: 8998, finalized_epoch: 8961, finalized_root: 0x655b…27d7, peers: 13, service: slot_notifier
Sep 13 12:53:25.997 INFO Synced                                  slot: 287966, block:    …  empty, epoch: 8998, finalized_epoch: 8961, finalized_root: 0x655b…27d7, peers: 12, service: slot_notifier
Sep 13 12:53:35.227 INFO Sync state updated                      new_state: Syncing Finalized Chain, old_state: Synced, service: sync
Sep 13 12:53:37.994 INFO Syncing                                 est_time: --, distance: 1120 slots (3 hrs 44 mins), peers: 14, service: slot_notifier
Sep 13 12:53:49.998 INFO Syncing                                 est_time: --, distance: 1121 slots (3 hrs 44 mins), peers: 14, service: slot_notifier
Sep 13 12:53:52.622 WARN Batch failed to download. Dropping chain scoring peers, batch_epoch: 8961, score_adjustment: Low Tolerance Error, chain: 5973664732859605117, service: sync
Sep 13 12:53:52.622 INFO Sync state updated                      new_state: Syncing Head Chain, old_state: Syncing Finalized Chain, service: sync
Sep 13 12:53:57.006 INFO Sync state updated                      new_state: Syncing Finalized Chain, old_state: Syncing Head Chain, service: sync
Sep 13 12:54:01.994 INFO Syncing                                 est_time: --, distance: 1122 slots (3 hrs 44 mins), peers: 18, service: slot_notifier
Sep 13 12:54:08.768 WARN Batch failed to download. Dropping chain scoring peers, batch_epoch: 8961, score_adjustment: Low Tolerance Error, chain: 2264951190402208954, service: sync
Sep 13 12:54:08.769 INFO Sync state updated                      new_state: Syncing Head Chain, old_state: Syncing Finalized Chain, service: sync
Sep 13 12:54:13.996 INFO Syncing                                 est_time: --, distance: 1123 slots (3 hrs 44 mins), peers: 18, service: slot_notifier
Sep 13 12:54:22.187 INFO Sync state updated                      new_state: Synced, old_state: Syncing Head Chain, service: sync
Sep 13 12:54:25.994 INFO Synced                                  slot: 287971, block:    …  empty, epoch: 8999, finalized_epoch: 8961, finalized_root: 0x655b…27d7, peers: 17, service: slot_notifier
Sep 13 12:54:32.165 INFO Sync state updated                      new_state: Syncing Finalized Chain, old_state: Synced, service: sync
Sep 13 12:54:37.992 INFO Syncing                                 est_time: --, distance: 1125 slots (3 hrs 45 mins), peers: 25, service: slot_notifier
```

## Expected Behaviour

How _should_ the application behave?

The beacon node should sync.

## Steps to resolve

Please describe the steps required to resolve this issue, if known.

Unknown.  I tried for a few hours to get the beacon node to sync before purging the DB and starting over.

**Comment by AgeManning** [comment]
This looks potentially like a database error. 

Can you post some debug logs to find out. 

You can run `lighthouse --debug-level debug bn ...` to get debug logs. 

I suspect this is a db error that occurred a few versions back. The DB needed to be re-downloaded to fully be resolved. 

I suspect doing doing a complete re-sync will resolve this. You can delete your db and start a resync from scratch by running:
```
lighthouse bn --purge-db ... 
```

**Comment by SimplyCorey** [comment]
Thanks for getting back to me.

This error can not be caused by a bug that was present in a previous version.  This node has only ever run the latest version of lighthouse. It has never been upgraded. Unfortunately, I can't get the debug logs at this point as I already purged the database and re-synced it.  

I will boot up another beacon node with debugging logs and try to make this happen again by forcing the container to stop repeatedly.

**Comment by AgeManning** [comment]
This should be resolved in #1629 

My assumption that a previous db error was causing this was invalid, in fact the issue was in fork choice.

**Comment by AgeManning** [comment]
There is still an issue with this we are looking into it. Will re-open this until its correctly resolved.
